common/beatmap_utils.py

```diff
@@ -0,0 +1,472 @@
+    elif ranked_status is RankedStatus.LOVED:
+        # loved maps can *technically* be updated
+        update_interval = timedelta(days=1)
+    elif ranked_status in (RankedStatus.RANKED, RankedStatus.APPROVED):
+        # in very rare cases, the osu! team has updated ranked/appvoed maps
+        # this is usually done to remove things like inappropriate content
+        update_interval = timedelta(days=1)
+    else:
+        raise NotImplementedError(f"Unknown ranked status: {ranked_status}")
+
+    return last_updated <= (datetime.now() - update_interval)
+
+
+@dataclass
+class Beatmap:
+    md5: str
+    id: int
+    set_id: int
+
+    song_name: str
+
+    status: RankedStatus
+
+    plays: int
+    passes: int
+    mode: int  # vanilla mode
+
+    od: float
+    ar: float
+
+    hit_length: int
+
+    last_update: int = 0
+
+    max_combo: int = 0
+    bpm: int = 0
+    filename: str = ""
+    frozen: bool = False
+    rating: Optional[float] = None
+
+    @property
+    def gives_pp(self) -> bool:
+        return self.status in (RankedStatus.RANKED, RankedStatus.APPROVED)
+
+    @property
+    def has_leaderboard(self) -> bool:
+        return self.status >= RankedStatus.RANKED
+
+    @property
+    def deserves_update(self) -> bool:
+        """Checks if there should be an attempt to update a map/check if
+        should be updated."""
+
+        return _should_get_updates(
+            int(self.status),
+            datetime.fromtimestamp(self.last_update),
+        )
+
+    def osu_string(self, score_count: int, rating: float) -> str:
+        return (
+            f"{int(self.status)}|false|{self.id}|{self.set_id}|{score_count}|0|\n"  # |0| = featured artist bs
+            f"0\n{self.song_name}\n{rating:.1f}"  # 0 = offset
+        )
+
+    def to_dict(self) -> dict[str, Any]:
+        return {
+            "beatmap_md5": self.md5,
+            "beatmap_id": self.id,
+            "beatmapset_id": self.set_id,
+            "song_name": self.song_name,
+            "ranked": self.status.value,
+            "playcount": self.plays,
+            "passcount": self.passes,
+            "mode": self.mode,
+            "od": self.od,
+            "ar": self.ar,
+            "hit_length": self.hit_length,
+            "latest_update": self.last_update,
+            "max_combo": self.max_combo,
+            "bpm": self.bpm,
+            "file_name": self.filename,
+            "ranked_status_freezed": self.frozen,
+            "rating": self.rating,
+        }
+
+    @classmethod
+    def from_mapping(cls, mapping: Mapping[str, Any]) -> Beatmap:
+        return cls(
+            md5=mapping["beatmap_md5"],
+            id=mapping["beatmap_id"],
+            set_id=mapping["beatmapset_id"],
+            song_name=mapping["song_name"],
+            status=RankedStatus(mapping["ranked"]),
+            plays=mapping["playcount"],
+            passes=mapping["passcount"],
+            mode=mapping["mode"],
```

Comment:
mode is now int instead of Mode, should be fine as bmaps are always stored under vanilla mode iirc?